tests/unit/test_charm.py

```diff
@@ -150,3 +155,57 @@ def test__loki_config(self):
                 sorted(logger.output),
                 ["DEBUG:charm:Loki Provider is available. Loki version: 3.14159"],
             )
+
+
+class TestDelayedPebbleReady(unittest.TestCase):
+    """Feature: Charm code must be resilient to any (reasonable) order of startup event firing."""
+
+    @patch_network_get(private_address="1.1.1.1")
+    @patch("charm.KubernetesServicePatch", lambda x, y: None)
+    @patch(
+        "charms.loki_k8s.v0.loki_push_api.LokiPushApiProvider._check_alert_rules", new=tautology
+    )
+    def setUp(self):
+        self.harness = Harness(LokiOperatorCharm)
+
+        # GIVEN this unit is the leader
+        self.harness.set_leader(True)
+
+        # AND a "logging" app joins with several units
+        self.log_rel_id = self.harness.add_relation("logging", "consumer-app")
+        self.harness.add_relation_unit(self.log_rel_id, "consumer-app/0")
+        self.harness.add_relation_unit(self.log_rel_id, "consumer-app/1")
+        self.harness.begin_with_initial_hooks()
+        # self.harness.charm.loki_provider._check_alert_rules = tautology
```

Comment:
@rbarry82 may I get a patching hint?
The decorator doesn't work for me here and I do not want to monkeypatch directly.

Reply:
The gist is that `@patch` only applies to the method it patches. In the case of `setUp`, it patches _only_ `setUp`, not each subsequent method.

So you can either also add the decorator to the entire class (just above `class TestDelayedPebbleReady`) *and* `setUp` (due to the object instantiation and realtime AST patching/parsing, it will apply to all methods _except_ `setUp` otherwise), or do it directly in `setUp` and stop in `tearDown`.

The latter is the "right" way to do it, and the rest of the mocks seem to have just sorta gotten lucky so far.

Reply:
I just pushed an update which demonstrates it.